File: app/src/main/java/com/example/acer/movies/adapters/RecyclerViewAdapterReviews.java
package com.example.acer.movies.adapters;

import android.content.Context;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.acer.movies.R;
import com.example.acer.movies.models.Review;

import java.util.ArrayList;

/**
 * Created by KeshavAggarwal on 14/03/17.
 */

public class RecyclerViewAdapterReviews extends RecyclerView.Adapter<RecyclerViewAdapterReviews.ViewHolder> {
    Context mContext;
    private ArrayList<Review> mReviews;
    private boolean isExpanded = false;

    public RecyclerViewAdapterReviews(ArrayList<Review> reviews, Context context) {
        mReviews = reviews;
        mContext = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.review_list_item_view, parent, false);
        return new ViewHolder(v);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        String secondInitialOfAuthor = null;
        for (int i = 0; i < mReviews.get(position).getAuthor().length(); i++) {
            if (mReviews.get(position).getAuthor().charAt(i) == ' ') {
                secondInitialOfAuthor = mReviews.get(position).getAuthor().substring(i + 1, i + 2);
                holder.circleTextView.setText(mReviews.get(position).getAuthor().substring(0, 1) + secondInitialOfAuthor);
                break;
            } else
                holder.circleTextView.setText(mReviews.get(position).getAuthor().substring(0, 1));
        }

        holder.reviewHeading.setText("A review by " + mReviews.get(position).getAuthor());
        holder.reviewSubheading.setText("By " + mReviews.get(position).getAuthor());
        holder.reviewTextView.setText(mReviews.get(position).getContent());

        holder.reviewTextView.post(new Runnable() {
            @Override
            public void run() {
                if (holder.reviewTextView.getLineCount() == 9) {
                    holder.readTheRestTextView.setVisibility(View.VISIBLE);
                    holder.readTheRestTextView.setText("Read More");
                    holder.readTheRestTextView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View v) {
                            if (isExpanded) {
                                holder.reviewTextView.setMaxLines(9);
                                holder.reviewTextView.setEllipsize(TextUtils.TruncateAt.END);
                                isExpanded = false;
                                holder.readTheRestTextView.setText("read more");


                            } else {
                                holder.reviewTextView.setMaxLines(Integer.MAX_VALUE);
                                holder.reviewTextView.setEllipsize(null);
                                isExpanded = true;
                                holder.readTheRestTextView.setText("hide");

                            }

                        }
                    });
                }

            }
        });

    }

    @Override
    public int getItemCount() {
        return mReviews.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        CardView reviewsCardView;
        TextView reviewHeading;
        TextView reviewSubheading;
        TextView reviewTextView;
        TextView circleTextView;
        TextView readTheRestTextView;

        public ViewHolder(View itemView) {
            super(itemView);

            reviewsCardView = (CardView) itemView.findViewById(R.id.reviewCardView);
            reviewTextView = (TextView) itemView.findViewById(R.id.reviewsTextView);
            reviewHeading = (TextView) itemView.findViewById(R.id.reviewHeading);
            circleTextView = (TextView) itemView.findViewById(R.id.circleTextView);
            reviewSubheading = (TextView) itemView.findViewById(R.id.reviewSubheading);
            readTheRestTextView = (TextView) itemView.findViewById(R.id.readtheRestTextView);
        }
    }

}

File: app/src/main/java/com/app/techvalley/movies/adapters/RecyclerViewAdapterReviews.kt
package com.app.techvalley.movies.adapters

import android.content.Context
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.app.techvalley.movies.R
import com.app.techvalley.movies.models.Review
import kotlinx.android.synthetic.main.review_list_item_view.view.*

import java.util.ArrayList

/**
 * Created by KeshavAggarwal on 14/03/17.
 */

class RecyclerViewAdapterReviews(private val mReviews: ArrayList<Review>, internal var mContext: Context) : RecyclerView.Adapter<RecyclerViewAdapterReviews.ViewHolder>() {
    private var isExpanded = false

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.review_list_item_view, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        var secondInitialOfAuthor: String? = null
        for (i in 0 until mReviews[position].author.length) {
            if (mReviews[position].author[i] == ' ') {
                secondInitialOfAuthor = mReviews[position].author.substring(i + 1, i + 2)
                holder.itemView.circleTV.text = mReviews[position].author.substring(0, 1) + secondInitialOfAuthor
                break
            } else
                holder.itemView.circleTV.text = mReviews[position].author.substring(0, 1)
        }
        holder.itemView.reviewHeadingTV.text = "A review by " + mReviews[position].author
        holder.itemView.reviewSubheadingTV.text = "By " + mReviews[position].author
        holder.itemView.reviewsTextView.text = mReviews[position].content

        holder.itemView.reviewsTextView.post {
            if (holder.itemView.reviewsTextView.lineCount == 9) {
                holder.itemView.readMoreTV.visibility = View.VISIBLE
                holder.itemView.readMoreTV.text = "Read More"
                holder.itemView.readMoreTV.setOnClickListener {
                    if (isExpanded) {
                        holder.itemView.reviewsTextView.maxLines = 9
                        holder.itemView.reviewsTextView.ellipsize = TextUtils.TruncateAt.END
                        isExpanded = false
                        holder.itemView.readMoreTV.text = "read more"
                    } else {
                        holder.itemView.reviewsTextView.maxLines = Integer.MAX_VALUE
                        holder.itemView.reviewsTextView.ellipsize = null
                        isExpanded = true
                        holder.itemView.readMoreTV.text = "hide"

                    }
                }
            }
        }

    }

    override fun getItemCount(): Int {
        return mReviews.size
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

}


